### PR TITLE
GPU Patch for Pytorch

### PR DESCRIPTION
**Pull request by danich1**
This quick patch adds GPU functionality towards the pytorch classifiers (LSTM and LogisticRegression). This patch provides a quick work around for issues #1023 and #974. I know @jdunnmon was working on a patch  (from reading the comments) and if he has finished feel free to close this PR. Otherwise feel free to incorporate this implementation.

### CODE REVIEW

**Comment by ajratner** [comment]
This looks great, thanks!!  @jdunnmon?

**Comment by danich1** [comment]
Um I think he was a post doc in the lab. (Tag was based off of #974) My bad if this is incorrect. Anyway... glad to get approval on this.

**Comment by ajratner** [comment]
Oh sorry :) . I was just tagging him so he can take a look- I know @jdunnmon very well :D !

File: snorkel/learning/pytorch/rnn/lstm.py [hunk 1]
```diff
@@ -43,9 +43,16 @@ def forward(self, X, hidden_state):
         seq_lengths, perm_idx = seq_lengths.sort(0, descending=True)
         X = X[perm_idx, :]
         inv_perm_idx = torch.tensor([i for i, _ in sorted(enumerate(perm_idx), key=lambda idx: idx[1])], dtype=torch.long)
-
         encoded_X = self.embedding(X)
-        encoded_X = pack_padded_sequence(encoded_X, seq_lengths, batch_first=True)
+
+        if self.gpu:
+            hidden_state = (hidden_state[0].permute(1,0,2), hidden_state[1].permute(1,0,2))
```
Comment (jdunnmon) [inline snorkel/learning/pytorch/rnn/lstm.py]: @danich1 why the permutation here (and below)?  Add a comment perhaps?

File: snorkel/learning/pytorch/rnn/lstm.py [hunk 2]
```diff
@@ -55,4 +62,7 @@ def initialize_hidden_state(self, batch_size):
         return (
             torch.zeros(self.num_layers * self.num_directions, batch_size, self.hidden_dim),
             torch.zeros(self.num_layers * self.num_directions, batch_size, self.hidden_dim)
+        ) if self.gpu else (
```
Comment (jdunnmon) [inline snorkel/learning/pytorch/rnn/lstm.py]: @danich1 seems like this condition might be the wrong way around?  I.e. `.cuda()` should happen if self.gpu is true, not the reverse (which is how it seems to be written now)?

@ajratner do you want to put in CI tests for GPU functionality before merging or no?

**Comment by jdunnmon** [comment]
@danich1 thanks for submitting the patch! just a couple of quick questions (see review)...

File: snorkel/learning/pytorch/rnn/lstm.py (hunk 2 above)
Comment (danich1) [reply snorkel/learning/pytorch/rnn/lstm.py]: Ack. Good catch. Changing that now.

File: snorkel/learning/pytorch/rnn/lstm.py (hunk 1 above)
Comment (danich1) [reply snorkel/learning/pytorch/rnn/lstm.py]: Long story short this is a trick to get dataparallel to work with the hidden state and cell state tensors. The dataparallel functions splits on the first dimension, which is implied to be the batch size dimension. This is correct for the input, but for the LSTM module it is a problem. The module expects both tensors to be in the form of ```num layers* num directions, batch_size, dimensions```, which will cause data parallel to split the data on the wrong dimension. (eventually leading to an error) The easiest solution is to permute the hidden state and cell state before the forward call and then permute it back in the foward function. More explanation in this [forum post](https://discuss.pytorch.org/t/multi-layer-rnn-with-dataparallel/4450/5).

**Comment by jdunnmon** [comment]
@danich1 thanks for the changes.  I was trying to see if there was a way to write tests for this using Travis, but I'm not having a lot of luck.  For now, can you put together a short notebook that should run on one of the tutorial datasets using your GPU patch?  That way I can run on my end and confirm it works before we merge.  @ajratner any other thoughts?

**Comment by danich1** [comment]
Yeah sure no problem. I'll upload a notebook shortly.

**Comment by ajratner** [comment]
@jdunnmon any updates here?

**Comment by jdunnmon** [comment]
@ajratner I've spent a bit testing this, and there are still issues.  Specifically, the `.cuda()` data types are not handled adequately throughout the code.  I've fixed a couple of small errors on my `jdunnmon:gpu` branch, but we still need to have appropriate checks for `cuda` datatypes.  

@danich1 any updates here?  have you been able to run this on the gpu on your side?

**Comment by danich1** [comment]
So far no updates. Been using metal framework, but I'll take a look at this issue when I get the chance.

**Comment by henryre** [comment]
Hi @danich1, we've moved the information extraction version of Snorkel (v0.7) to https://github.com/snorkel-team/snorkel-extraction. Feel free to reopen this PR there! The new Snorkel [MultitaskClassifier](https://snorkel.readthedocs.io/en/v0.9.0/packages/_autosummary/classification/snorkel.classification.MultitaskClassifier.html#snorkel.classification.MultitaskClassifier) makes it much easier to work with PyTorch GPU!

**Pull request closed without merging**